frame/elections-phragmen/src/lib.rs

```diff
@@ -466,34 +634,73 @@ decl_module! {
 		/// Note that this does not affect the designated block number of the next election.
 		///
 		/// # <weight>
-		/// #### State
-		/// Reads: O(do_phragmen)
-		/// Writes: O(do_phragmen)
+		/// If we have a replacement:
+		/// 	- Base weight: 47.15 us
+		/// 	- State reads:
+		/// 		- RunnersUp.len()
+		/// 		- Members, RunnersUp (remove_and_replace_member)
+		/// 	- State writes:
+		/// 		- Members, RunnersUp (remove_and_replace_member)
+		/// Else, since this is a root call and will go into phragmen, we assume full block for now.
 		/// # </weight>
-		#[weight = (2_000_000_000, DispatchClass::Operational)]
-		fn remove_member(origin, who: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+		#[weight = FunctionOf(
+			|(_, has_replacement): (_, &bool)| if *has_replacement {
+				48 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(3, 2)
+			} else {
+				T::MaximumBlockWeight::get()
+			},
+			DispatchClass::Normal,
+			Pays::Yes,
+		)]
+		fn remove_member(
+			origin,
+			who: <T::Lookup as StaticLookup>::Source,
+			has_replacement: bool,
+		) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
 			let who = T::Lookup::lookup(who)?;
 
+			let will_have_replacement = <RunnersUp<T>>::decode_len().unwrap_or(0) > 0;
+			match (will_have_replacement, has_replacement) {
+				(true, true) | (false, false) => {
+					// correct prediction. nothing
+				}
+				(false, true) => {
+					// prediction was that we will have a replacement, so we don't charge a whole
+					// lot of weight. just abort.
+					return Err(Error::<T>::InvalidReplacement.into());
+				}
+				(true, false) => {
+					// prediction was that we will NOT have a replacement, and now this call is
+					// aborting whilst charging a metric ton of weight. Refund and abort.
+					return Err(Error::<T>::InvalidReplacement.with_weight(
+						// refund to the variant where we have a replacement. This still an
+						// overestimate but fine for now.
+						48 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(1, 0)
+					).into());
```

Comment:
You can lose that `.into()`. `with_weight` already returns the correct type AFAIK.